### PR TITLE
fix: pushing behavior

### PR DESCRIPTION
**Pull request by kelvin-olaiya**
The `PhysicalPedestrian2D` class was extending `PhysicalPedestrian` which implemented `physicalForces` function by using an adimensional shape to compute forces. 
Overriding the `comfortArea` property in `PhysicalPedestrian2D` was not enough to compute `physicalForces` with a 2d-shape. So i made `PhysicalPedestrian` abstract and implemented the function directly in `PhysicalPedestrian2D`. 
I then added a test for pushing behavior.

### CODE REVIEW

File: alchemist-physical-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/PhysicalPedestrian.kt [hunk 1]
```diff
@@ -27,15 +27,15 @@ import it.unibo.alchemist.model.interfaces.Node.Companion.asPropertyOrNull
 /**
  * Base implementation of a pedestrian's capability to experience physical interactions.
  */
-class PhysicalPedestrian<T, P, A, F>(
+abstract class PhysicalPedestrian<T, P, A, F>(
```
Comment (DanySK) [inline alchemist-physical-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/PhysicalPedestrian.kt:30]: Whi is this called "pedestrian"? Isn't this just a node with some physics?

File: alchemist-physical-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/PhysicalPedestrian2D.kt [hunk 2]
```diff
@@ -25,15 +26,34 @@ import org.apache.commons.math3.random.RandomGenerator
...
-        .shapeFactory.circle(node.asProperty<T, AreaProperty<T>>().shape.radius + comfortRay)
+        .shapeFactory
+        .circle(node.asProperty<T, AreaProperty<T>>().shape.radius + comfortRay)
+        .transformed { origin(environment.getPosition(node)) }
+
+    override fun physicalForces(
+        environment: PhysicsEnvironment<T, Euclidean2DPosition, Euclidean2DTransformation, Euclidean2DShapeFactory>,
+    ) = environment.getNodesWithin(comfortArea)
+        .minusElement(node)
+        .filter { it.asPropertyOrNull<T, AreaProperty<T>>() != null }
```
Comment (DanySK) [inline alchemist-physical-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/PhysicalPedestrian2D.kt:51]:
```suggestion
        .filterNotNull { it.asPropertyOrNull<T, AreaProperty<T>> }
```

File: alchemist-physical-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/PhysicalPedestrian2D.kt [hunk 3]
```diff
@@ -25,15 +26,34 @@ import org.apache.commons.math3.random.RandomGenerator
...
+
     override val comfortArea: Euclidean2DShape get() = environment
-        .shapeFactory.circle(node.asProperty<T, AreaProperty<T>>().shape.radius + comfortRay)
+        .shapeFactory
+        .circle(node.asProperty<T, AreaProperty<T>>().shape.radius + comfortRay)
+        .transformed { origin(environment.getPosition(node)) }
+
+    override fun physicalForces(
+        environment: PhysicsEnvironment<T, Euclidean2DPosition, Euclidean2DTransformation, Euclidean2DShapeFactory>,
+    ) = environment.getNodesWithin(comfortArea)
```
Comment (DanySK) [inline alchemist-physical-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/PhysicalPedestrian2D.kt:49]: use sequences, there are too many chained calls for eager collections

File: alchemist-physical-agents/src/main/kotlin/it/unibo/alchemist/model/interfaces/properties/PhysicalPedestrian.kt [hunk 4]
```diff
@@ -47,5 +47,5 @@ interface PhysicalPedestrian<T, P, A, F> : PhysicalProperty<T, P, A, F>
 /**
  * A pedestrian's capability to experience physical forces in a 2D space.
  */
-interface PhysicalPedestrian2DProperty<T> :
+interface PhysicalPedestrian2D<T> :
```
Comment (DanySK) [inline alchemist-physical-agents/src/main/kotlin/it/unibo/alchemist/model/interfaces/properties/PhysicalPedestrian.kt:50]: Why pedestrian?

File: alchemist-physical-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/PhysicalPedestrian.kt [hunk 5]
```diff
@@ -53,9 +53,13 @@ class PhysicalPedestrian<T, P, A, F>(
...
     }
 
+    private fun getShapeInNodePosition(node: Node<T>) = node.asProperty<T, OccupiesSpaceProperty<T, P, A>>()
+        .shape
+        .transformed { origin(environment.getPosition(node)) }
+
     override fun repulsionForce(other: Node<T>): P {
-        val myShape = node.asProperty<T, OccupiesSpaceProperty<T, P, A>>().shape
-        val otherShape = other.asProperty<T, OccupiesSpaceProperty<T, P, A>>().shape
+        val myShape = getShapeInNodePosition(node)
```
Comment (DanySK) [inline alchemist-physical-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/PhysicalPedestrian.kt:61]: Save the local node shape property upfront in the constructor

File: alchemist-physical-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/actions/physicalstrategies/Sum.kt [hunk 6]
```diff
@@ -29,6 +29,6 @@ class Sum<T>(
 ) : PhysicalSteeringStrategy<T, Euclidean2DPosition, Euclidean2DTransformation, Euclidean2DShapeFactory> {
 
     override fun computeNextPosition(overallIntentionalForce: Euclidean2DPosition): Euclidean2DPosition =
-        (node.asProperty<T, PhysicalPedestrian2DProperty<T>>().physicalForces(environment) + overallIntentionalForce)
+        (node.asProperty<T, PhysicalPedestrian2D<T>>().physicalForces(environment) + overallIntentionalForce)
```
Comment (DanySK) [inline alchemist-physical-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/actions/physicalstrategies/Sum.kt:32]: nope. Save the property in the constructor, there is no need to recompute it over and over

File: alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/Social.kt [hunk 7]
```diff
@@ -24,4 +24,6 @@ data class Social<T> @JvmOverloads constructor(
 ) : SocialProperty<T> {
 
     override fun cloneOnNewNode(node: Node<T>) = Social(node, GenericGroup(listOf(node)))
+
+    override fun toString(): String = group.toString()
```
Comment (DanySK) [inline alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/Social.kt:28]:
```suggestion
    override fun toString(): String = "${this::class.simpleName}$group"
```

**Review by DanySK — CHANGES_REQUESTED** [review]

File: alchemist-physical-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/PhysicalPedestrian2D.kt (hunk 2 above)
Comment (kelvin-olaiya) [reply alchemist-physical-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/PhysicalPedestrian2D.kt:51]: `filterNotNull` takes no arguments and filters non-null elements in the pipeline

File: alchemist-physical-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/PhysicalPedestrian.kt (hunk 1 above)
Comment (kelvin-olaiya) [reply alchemist-physical-agents/src/main/kotlin/it/unibo/alchemist/model/implementations/properties/PhysicalPedestrian.kt:30]: It captures elements such as *comfortArea* and *comfortRay* which i though are typical of a pedestrian.
However, there is a more general interface called `PhysicalProperty`.

**Comment by DanySK** [comment]
Let me know when it's ready

**Comment by kelvin-olaiya** [comment]
It's ready. Sorry, I thought I've sent the review request

**Review by DanySK — APPROVED** [review]

**Pull request merged**